runtime/src/accounts_background_service.rs

```diff
@@ -169,23 +174,60 @@ impl SnapshotRequestHandler {
                     shrink_time.stop();
                 }
 
+                let block_height = snapshot_root_bank.block_height();
+                let snapshot_type = if block_height
+                    % self.snapshot_config.full_snapshot_archive_interval_slots
+                    == 0
+                {
+                    Some(SnapshotType::FullSnapshot)
+                } else if block_height
+                    % self
+                        .snapshot_config
+                        .incremental_snapshot_archive_interval_slots
+                    == 0
+                {
+                    Some(SnapshotType::IncrementalSnapshot(last_full_snapshot_slot))
+                } else {
+                    None
+                };
+
                 // Generate an accounts package
+                let mut attempts = 0;
                 let mut snapshot_time = Measure::start("snapshot_time");
-                let r = snapshot_utils::snapshot_bank(
-                    &snapshot_root_bank,
-                    status_cache_slot_deltas,
-                    &self.accounts_package_sender,
-                    &self.snapshot_config.bank_snapshots_dir,
-                    &self.snapshot_config.snapshot_archives_dir,
-                    self.snapshot_config.snapshot_version,
-                    &self.snapshot_config.archive_format,
-                    hash_for_testing,
-                );
-                if r.is_err() {
+                loop {
+                    attempts += 1;
+                    let result = snapshot_utils::snapshot_bank(
+                        &snapshot_root_bank,
+                        status_cache_slot_deltas.clone(),
+                        &self.accounts_package_sender,
+                        &self.snapshot_config.bank_snapshots_dir,
+                        &self.snapshot_config.snapshot_archives_dir,
+                        self.snapshot_config.snapshot_version,
+                        self.snapshot_config.archive_format,
+                        hash_for_testing,
+                        snapshot_type,
+                    );
+
+                    if result.is_ok() {
+                        break;
```

Comment:
`snapshot_bank` can throw an error on channel send as well, if the thread on the receiving side of the `accounts_package_sender`  closes first. This is possible behavior when the validator is shutting down, so some options:

1. maybe we need check if result `is_ok` OR it's an expected send error, then break.

2. Eat and don't the send error at all: https://github.com/solana-labs/solana/blob/6d246031d06b81ee21d591a267a50f9f76d93eac/runtime/src/snapshot_utils.rs#L1579 with something like `let _ = accounts_package_sender.send(accounts_package)`, 

3. Have the `snapshot_bank` handle some input `max_retries` and wrap that logic in there.

I think 1 is cleanest, but is annoying to maintain because if we add additional errors in `snapshot_bank` we could run into this again. Maybe 1) AND we define some `enum SnapshotBankErrors` returned by `snapshot_bank` such that if we introduce any new errors, we'll have to handle the new member in the enum

Reply:
Carl and I talked offline. Retrying in ABS has been removed. Filesystem errors shouldn't happen, so any fs:: calls now are `expect()` instead of `?`.